content/news/2022-01-06-bevy-0.6/index.md

```diff
@@ -0,0 +1,1007 @@
+
+```rust
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, AppLabel)]
+pub struct RenderApp;
+
+let mut render_app = App::empty();
+app.add_sub_app(RenderApp, render_app, move |app_world, render_app| {
+    // execute app logic here
+});
+
+// later
+app.sub_app_mut(RenderApp)
+    .add_system(some_system);
+    .add_system(some_other_system);
+```
+
+We plan on exposing more control over scheduling and running sub apps in the future.
+
+### Query::iter_combinations
+
+<div class="release-feature-authors">authors: @Frizi</div>
+
+You can now iterate all permutations of N entities for a given query:
+
+```rust
+fn system(query: Query<&Player>) {
+    // iterates each possible combination of two entities exactly once
+    for [e1, e2] in query.iter_combinations() {
+    }
+
+    // iterates each possible combination of three entities exactly once
+    for [e1, e2, e3] in query.iter_combinations() {
+    }
+}
+```
+
+This is especially useful for things like "checking for entities for collisions with all other entities". There is also an `iter_combinations_mut` variant. Just be careful ... the time complexity of this grows quickly with large numbers of combinations. With great power comes great responsibility!
+
+The new `iter_combinations.rs` example illustrats how to use this new API to calculate gravity between objects in a "solar system":
+
+![iter_combinations](iter_combinations.png)
+
+### Optimized System Commands
+
+<div class="release-feature-authors">authors: @NathanSWard</div>
+
+System Commands got a nice performance boost by changing how command buffers are stored and reused:
+
+#### Entity Spawn Benchmark Duration (in microseconds, less is better)
+
+This benchmark spawns entities with a variety of component compositions to ensure we cover a variety of cases. Treat these numbers as realtives, not absolutes.
+
+![commands perf](commands_perf.svg)
+
+### System Param Lifetimes
+
+<div class="release-feature-authors">authors: @cart, @BoxyUwU, @TheRawMeatball</div>
+
+System and Query lifetimes were made more explicit by splitting out the `'system` and `'world`, lifetimes and using them explicitly where possible. This enables Rust to reason about ECS lifetimes more effectively, especially for read-only lifetimes. This was particularly important because it enabled the new Bevy Renderer to convince WGPU that ECS resources actually live for as long as the Render World.
+
+### Soundness / Correctness Improvements
+
+<div class="release-feature-authors">authors: @BoxyUwU, @TheRawMeatball, @Frizi, @thebluefish, @sapir, @bjorn3, @DJMcNab</div>
+
+Bevy ECS received a solid number of soundness and correctness improvements this release, alongside some unsafe code removals. Queries and internal storages like Tables and BlobVecs in particular had a number of improvements in these areas. And we have even more improvements in the works!
```

Comment:
Agreed, it's not a good look

I'd be tempted to remove this section entirely, or least not mentioning the soundness fixes.

Perhaps:

We've been working on reducing and scoping the unsafe code in our core ECS abstraction, to make bevy_ecs's correctness easier to reason about. Work on this will continue into the future.

Reply:
No. I don't think we should be worried about "a good look" or "saving face". We should be honest, open, and transparent. We had bugs -- and we owe it to the community to clearly communicate about it. Not great to pretend that everything is rosy.

I stand by what I said. We should say we found soundness bugs and fixed them. We should say that we take code safety seriously and will continue with our efforts to improve in this direction.

Reply:
Yup we definitely need to cover soundness/safety work (both because the people doing the work deserve the credit and its worth being open about this).

However I do feel the need to soften the language here _precisely_ because people take this so seriously. A portion of people who blindly "hate everything with unsafe" will write us off if we aren't careful. We don't have the space (or user attention span) to properly argue points about _why_ Bevy ECS requires unsafe code to provide the features that it does, why some level of risk is necessary to properly explore the design space, the project's (and my own) progression from "use unsafe without proper scrutiny" to "cut it out as much as possible", etc, etc. This is a "win hearts and minds" blog post. I don't think we can accomplish that goal if we broach this subject with a greater level of detail than we already have. I'm happy to publicly cover this subject with complete transparency in a different public setting where we can give it the nuance it deserves.

But I do agree that this could use some rephrasing to avoid conflating "resolved safety bug / mistake" with "safety improvement as a feature" and properly stress that we take safety issues seriously. That is a bad look. I'll rework this a bit.

Reply:
Just rephrased things a bit.